zebra-scan/src/tests.rs

```diff
@@ -166,6 +168,118 @@ async fn scanning_from_fake_generated_blocks() -> Result<()> {
+    /// The extended Sapling viewing key of [ZECpages](https://zecpages.com/boardinfo)
+    const ZECPAGES_VIEWING_KEY: &str = "zxviews1q0duytgcqqqqpqre26wkl45gvwwwd706xw608hucmvfalr759ejwf7qshjf5r9aa7323zulvz6plhttp5mltqcgs9t039cx2d09mgq05ts63n8u35hyv6h9nc9ctqqtue2u7cer2mqegunuulq2luhq3ywjcz35yyljewa4mgkgjzyfwh6fr6jd0dzd44ghk0nxdv2hnv4j5nxfwv24rwdmgllhe0p8568sgqt9ckt02v2kxf5ahtql6s0ltjpkckw8gtymxtxuu9gcr0swvz";
+
+    // Parse the key from ZECpages
+    let efvk = decode_extended_full_viewing_key(
+        HRP_SAPLING_EXTENDED_FULL_VIEWING_KEY,
+        ZECPAGES_VIEWING_KEY,
+    )
+    .unwrap();
+
+    let account = AccountId::from(1);
+
+    // Build a vector of viewing keys `vks` to scan for.
+    let fvk = efvk.fvk;
+    let ivk = fvk.vk.ivk();
+    let vks: Vec<(&AccountId, &SaplingIvk)> = vec![(&account, &ivk)];
+
+    let network = zebra_chain::parameters::Network::default();
+
+    // Create a continuous chain of mainnet blocks from genesis
+    let blocks: Vec<Arc<Block>> = zebra_test::vectors::CONTINUOUS_MAINNET_BLOCKS
+        .iter()
+        .map(|(_height, block_bytes)| block_bytes.zcash_deserialize_into().unwrap())
+        .collect();
+
+    // Create a populated state service.
+    let (_state_service, read_only_state_service, latest_chain_tip, _chain_tip_change) =
+        zebra_state::populated_state(blocks.clone(), network).await;
+
+    let db = read_only_state_service.db();
+
+    // use the tip as starting height
+    let mut height = latest_chain_tip.best_tip_height().unwrap();
+
+    let mut transactions_found = 0;
+    let mut transactions_scanned = 0;
+    let mut blocks_scanned = 0;
+    while let Some(block) = db.block(height.into()) {
+        // TODO: fix this issue in zcash_client_backend
+        let sapling_tree_size = if height.is_min() {
+            1
+        } else {
+            db.sapling_tree_by_hash_or_height(height.into())
+                .expect("should exist")
+                .count()
+        };
+
+        let orchard_tree_size = db
+            .orchard_tree_by_hash_or_height(height.into())
+            .expect("should exist")
+            .count();
+
+        let chain_metadata = ChainMetadata {
+            sapling_commitment_tree_size: sapling_tree_size
+                .try_into()
+                .expect("position should fit in u32"),
+            orchard_commitment_tree_size: orchard_tree_size
+                .try_into()
+                .expect("position should fit in u32"),
+        };
+
+        let block_metadata = if height.is_min() {
+            None
+        } else {
+            Some(BlockMetadata::from_parts(
+                height.previous()?.0.into(),
+                BlockHash(block.header.previous_block_hash.0),
+                db.sapling_tree_by_hash_or_height(block.header.previous_block_hash.into())
+                    .expect("should exist")
+                    .count()
+                    .try_into()
+                    .expect("should fit in u32"),
```

Comment:
Nitpick: This is part of the protocol spec, so it's not a "should", it's a requirement. This also applies to every similar sapling and orchard message in this file.

Just FYI the sprout limit is `2^29`.

```
                    .expect("sapling position is limited to u32::MAX"),
```